File: Data Structures/Stacks & Queues/stack.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Stack:
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0

    def peek(self):
        return self.head

    def push(self, item):
        new_node = Node(item)

        if self.is_empty():
            self.top = new_node
            self.bottom = new_node
        else:
            new_node.next = self.top
            self.top = new_node

        self.length += 1

    def pop(self):
        if self.is_empty():
            return None

        if (self.length == 1):
            self.bottom = None

        holding_pointer = self.top
        self.top = holding_pointer.next

        self.length -= 1

        return holding_pointer

    def is_empty(self):
        return self.length == 0


my_stack = Stack()
my_stack.push("Part 1")
my_stack.push("Part 2")
my_stack.push("Part 3")
my_stack.push("Part 3b")
my_stack.pop()
print(my_stack.peek())

File: Data Structures/Stacks & Queues/stack.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor() {
    this.top = null;
    this.bottom = null;
    this.length = 0;
  }

  /**
   * @returns the element at the top
   */
  peek() {
    return this.top;
  }

  /**
   * add element to the top of the stack
   */
  push(element) {
    const newNode = new Node(element);
    if (this.isEmpty()) {
      this.top = newNode;
      this.bottom = newNode;
    } else {
      const holdingPointer = this.top;
      this.top = newNode;
      this.top.next = holdingPointer;
    }

    this.length++;
  }

  /**
   * removes the element at the top of the stack
   * @returns {any} the removed node
   */
  pop() {
    if (this.isEmpty()) return null;

    if (this.length === 1) {
      this.bottom = null;
    }

    const holdingPointer = this.top;
    this.top = this.top.next;
    this.length--;
    return holdingPointer;
  }

  /**
   * @returns {boolean}
   */
  isEmpty() {
    return this.length === 0;
  }
}

const myStack = new Stack();
myStack.push("Coding");
myStack.push("is");
myStack.push("really");
myStack.push("difficult");
myStack.pop();
myStack.push("easy");
console.dir(myStack, { depth: 10 });
